File: priv/combined_metrics/samples/code_smells/no_debug_print_statements/good/payment.rb
require 'logger'

class Payment
  def initialize
    @logger = Logger.new($stdout)
  end

  def charge(user, amount, card)
    validated = validate_card(card)
    unless validated
      @logger.warn("card validation failed user=#{user.id}")
      return nil
    end

    result = call_gateway(validated, amount)
    @logger.info("payment charged user=#{user.id} amount=#{amount}")
    result
  end

  def refund(transaction_id, amount)
    transaction = fetch_transaction(transaction_id)
    return { error: :not_found } unless transaction
    return { error: :exceeds_original } if transaction[:amount] < amount

    result = call_refund_api(transaction, amount)
    @logger.info("refund processed tx=#{transaction_id} amount=#{amount}")
    result
  end

  def calculate_fee(amount, method)
    case method
    when :credit_card then amount * 0.029 + 0.30
    when :debit_card then amount * 0.015
    when :bank_transfer then 0.25
    else amount * 0.035
    end
  end

  private

  def validate_card(card)
    card
  end

  def call_gateway(_card, _amount)
    { transaction_id: 'txn_123' }
  end

  def fetch_transaction(_id)
    { amount: 100.0 }
  end

  def call_refund_api(_transaction, _amount)
    { status: :refunded }
  end
end

File: priv/combined_metrics/samples/code_smells/no_debug_print_statements/good/payment.js
import { logger } from './logger.js';

export class Payment {
  charge(user, amount, card) {
    const validated = this.validateCard(card);
    if (!validated) {
      logger.warn('card validation failed', { userId: user.id });
      return null;
    }

    const result = this.callGateway(validated, amount);
    logger.info('payment charged', { userId: user.id, amount });
    return result;
  }

  refund(transactionId, amount) {
    const transaction = this.fetchTransaction(transactionId);
    if (!transaction) {
      return { error: 'not_found' };
    }

    if (transaction.amount < amount) {
      return { error: 'exceeds_original' };
    }

    const result = this.callRefundApi(transaction, amount);
    logger.info('refund processed', { transactionId, amount });
    return result;
  }

  calculateFee(amount, method) {
    switch (method) {
      case 'credit_card':
        return amount * 0.029 + 0.3;
      case 'debit_card':
        return amount * 0.015;
      case 'bank_transfer':
        return 0.25;
      default:
        return amount * 0.035;
    }
  }

  validateCard(card) {
    return card;
  }

  callGateway(_card, _amount) {
    return { transactionId: 'txn_123' };
  }

  fetchTransaction(_id) {
    return { amount: 100.0 };
  }

  callRefundApi(_transaction, _amount) {
    return { status: 'refunded' };
  }
}
